File: cross_dataset_analyses/scripts/05_housekeeping_control.R
# ============================================================================
# Task 1: Housekeeping Gene Negative Control
# Tests whether the activation-dependent retained-intron pattern at Klrk1 is
# gene-specific or a pipeline artifact (e.g. gDNA contamination).
#
# Hypothesis: Actb and Gapdh - which are constitutively transcribed but NOT
# subject to activation-dependent splicing regulation - should show
# negligible retained-intron isoform expression across all conditions.
# ============================================================================

library(tidyverse)

# cross-dataset control: paths from shared config
PROJECT_ROOT <- Sys.getenv("KLRK1_ROOT",
                  "/Users/study/Desktop/Karimi/Klrk1_GVHD_project")
source(file.path(PROJECT_ROOT, "shared", "scripts_common", "config.R"))
table_dir <- shared_dir("results")   # shared/cross_dataset_analyses/results
fig_dir   <- shared_dir("figures")   # shared/cross_dataset_analyses/figures

# ---- Transcript annotations (from GRCm39 Ensembl 111 cDNA FASTA headers) ----

actb_meta <- tibble(
  transcript_id = c(
    "ENSMUST00000100497", "ENSMUST00000167721", "ENSMUST00000171419",
    "ENSMUST00000163829", "ENSMUST00000106216",
    "ENSMUST00000196997", "ENSMUST00000167386",
    "ENSMUST00000165629", "ENSMUST00000164765"
  ),
  gene = "Actb",
  biotype = c(rep("protein_coding", 5),
              rep("protein_coding_CDS_not_defined", 2),
              rep("retained_intron", 2))
)

gapdh_meta <- tibble(
  transcript_id = c(
    "ENSMUST00000117757", "ENSMUST00000118875", "ENSMUST00000073605",
    "ENSMUST00000183272", "ENSMUST00000182052", "ENSMUST00000182277",
    "ENSMUST00000144205", "ENSMUST00000182115", "ENSMUST00000182670",
    "ENSMUST00000192506", "ENSMUST00000147954", "ENSMUST00000182464",
    "ENSMUST00000144588"
  ),
  gene = "Gapdh",
  biotype = c(rep("protein_coding", 6),
              rep("protein_coding_CDS_not_defined", 3),
              rep("retained_intron", 4))
)

klrk1_meta <- tibble(
  transcript_id = c(
    "ENSMUST00000032252", "ENSMUST00000095412", "ENSMUST00000168919",
    "ENSMUST00000152256", "ENSMUST00000137660", "ENSMUST00000204694"
  ),                                    # all SIX Klrk1 transcripts (Klrk1-202 now included)
  gene = "Klrk1",
  biotype = c("protein_coding", "protein_coding", "protein_coding",
              "retained_intron", "retained_intron", "retained_intron")
)

all_meta <- bind_rows(actb_meta, gapdh_meta, klrk1_meta)
target_tx <- all_meta$transcript_id

# ---- All Salmon outputs across every dataset (datasets/*/data/salmon/*/quant.sf) ----
qfiles <- all_salmon_quants()
cat("Salmon quant.sf files found:", length(qfiles), "\n")

read_targets <- function(qpath) {
  if (!file.exists(qpath)) return(NULL)
  # sample name = the sample folder name (parent dir of quant.sf)
  sample_name <- basename(dirname(qpath))
  read_tsv(qpath, col_types = "cdddd", show_col_types = FALSE) %>%
    mutate(transcript_id = sub("\\.\\d+$", "", Name)) %>%
    filter(transcript_id %in% target_tx) %>%
    transmute(sample = sample_name, transcript_id, TPM)
}

all_tpm <- map_dfr(qfiles, read_targets) %>%
  left_join(all_meta, by = "transcript_id")

cat("Total measurements:", nrow(all_tpm), "\n")

# ---- Per-gene summary by sample ----
gene_totals <- all_tpm %>%
  group_by(sample, gene) %>%
  summarise(total_gene_tpm = sum(TPM), .groups = "drop")

ri_per_sample <- all_tpm %>%
  filter(biotype == "retained_intron") %>%
  group_by(sample, gene) %>%
  summarise(ri_tpm = sum(TPM), .groups = "drop") %>%
  left_join(gene_totals, by = c("sample", "gene")) %>%
  mutate(ri_pct = ifelse(total_gene_tpm > 0, ri_tpm / total_gene_tpm * 100, 0))

# ---- Master summary table ----
summary_tbl <- ri_per_sample %>%
  group_by(gene) %>%
  summarise(
    n_samples              = n(),
    mean_total_gene_tpm    = mean(total_gene_tpm),
    sd_total_gene_tpm      = sd(total_gene_tpm),
    mean_RI_tpm            = mean(ri_tpm),
    max_RI_tpm             = max(ri_tpm),
    n_samples_RI_above_1   = sum(ri_tpm > 1),
    n_samples_RI_above_5   = sum(ri_tpm > 5),
    mean_RI_pct            = mean(ri_pct),
    sd_RI_pct              = sd(ri_pct),
    .groups = "drop"
  ) %>%
  arrange(factor(gene, levels = c("Actb", "Gapdh", "Klrk1")))

cat("\n=== HOUSEKEEPING NEGATIVE CONTROL - SUMMARY ===\n")
print(summary_tbl, width = Inf)

write_csv(summary_tbl,    file.path(table_dir, "TableV2_housekeeping_summary.csv"))
write_csv(ri_per_sample,  file.path(table_dir, "TableV2_housekeeping_per_sample.csv"))
write_csv(all_tpm,        file.path(table_dir, "TableV2_housekeeping_all_transcripts.csv"))

# ---- Per-RI-transcript max TPM (the strict test) ----
ri_tx_max <- all_tpm %>%
  filter(biotype == "retained_intron") %>%
  group_by(gene, transcript_id) %>%
  summarise(
    max_TPM = max(TPM),
    mean_TPM = mean(TPM),
    n_samples_above_1_TPM = sum(TPM > 1),
    .groups = "drop"
  ) %>%
  arrange(gene, desc(max_TPM))

cat("\n=== PER-RI-TRANSCRIPT MAX EXPRESSION (across all samples) ===\n")
print(ri_tx_max, n = 30, width = Inf)

write_csv(ri_tx_max, file.path(table_dir, "TableV2_RI_transcript_max_per_gene.csv"))

# ---- Visualization ----
plot_data <- ri_per_sample %>%
  mutate(gene = factor(gene, levels = c("Actb", "Gapdh", "Klrk1")))

p <- ggplot(plot_data, aes(x = gene, y = ri_pct, fill = gene)) +
  geom_boxplot(width = 0.5, outlier.shape = NA, alpha = 0.7) +
  geom_jitter(width = 0.15, size = 1.4, alpha = 0.65) +
  scale_fill_manual(values = c("Actb" = "#7FBF7B", "Gapdh" = "#7FBF7B", "Klrk1" = "#D6604D")) +
  labs(
    title    = "Retained-Intron Isoform Proportion: Klrk1 vs Housekeeping Genes",
    subtitle = paste0("Each dot = one biological sample; bulk RNA-seq Salmon TPM (n=",
                      dplyr::n_distinct(all_tpm$sample), " samples total)"),
    x        = NULL,
    y        = "Retained-intron isoforms (% of total gene TPM)"
  ) +
  theme_classic(base_size = 11) +
  theme(legend.position = "none",
        plot.title = element_text(face = "bold", size = 12))

ggsave(file.path(fig_dir, "FigV5_housekeeping_negative_control.pdf"), p, width = 7, height = 5.5)
ggsave(file.path(fig_dir, "FigV5_housekeeping_negative_control.png"), p, width = 7, height = 5.5, dpi = 300)

cat("\n=== Outputs ===\n")
cat("Tables:", table_dir, "\n")
cat("Figure:", file.path(fig_dir, "FigV5_housekeeping_negative_control.png"), "\n")

File: python_pipeline/scripts/05_housekeeping_control.py
#!/usr/bin/env python3
# ============================================================================
# 05_housekeeping_control.py — Python port of 05_housekeeping_control.R
# Negative control: is the retained-intron (RI) proportion at Klrk1 far above the
# RI proportion at housekeeping genes Actb/Gapdh? (rules out gDNA contamination /
# pipeline artifact). Reads ALL Salmon outputs across datasets. Run: python3 scripts/05_*.py
# ============================================================================
import os, sys, glob
import pandas as pd, numpy as np
import matplotlib; matplotlib.use("Agg"); import matplotlib.pyplot as plt
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import config as cfg

# transcript -> (gene, is_retained_intron)   [from GRCm39 Ensembl cDNA headers]
ACTB_RI  = {"ENSMUST00000165629", "ENSMUST00000164765"}
ACTB_ALL = {"ENSMUST00000100497","ENSMUST00000167721","ENSMUST00000171419",
            "ENSMUST00000163829","ENSMUST00000106216","ENSMUST00000196997",
            "ENSMUST00000167386"} | ACTB_RI
GAPDH_RI  = {"ENSMUST00000192506","ENSMUST00000147954","ENSMUST00000182464","ENSMUST00000144588"}
GAPDH_ALL = {"ENSMUST00000117757","ENSMUST00000118875","ENSMUST00000073605",
             "ENSMUST00000183272","ENSMUST00000182052","ENSMUST00000182277",
             "ENSMUST00000144205","ENSMUST00000182115","ENSMUST00000182670"} | GAPDH_RI
KLRK1_ALL = set(cfg.KLRK1)
KLRK1_RI  = {t for t, b in cfg.BIOTYPE.items() if b == "retained_intron"}
GENES = {"Actb": (ACTB_ALL, ACTB_RI), "Gapdh": (GAPDH_ALL, GAPDH_RI),
         "Klrk1": (KLRK1_ALL, KLRK1_RI)}
ALL_TX = ACTB_ALL | GAPDH_ALL | KLRK1_ALL

quants = glob.glob(os.path.join(cfg.PROJECT_ROOT, "datasets", "*", "data", "salmon", "*", "quant.sf"))
print(f"Salmon quant.sf files found: {len(quants)}")

recs = []
for q in quants:
    sample = os.path.basename(os.path.dirname(q))
    df = pd.read_csv(q, sep="\t"); df["tx"] = df["Name"].str.replace(r"\.\d+$", "", regex=True)
    tpm = dict(zip(df["tx"], df["TPM"]))
    for gene, (allset, riset) in GENES.items():
        total = sum(tpm.get(t, 0) for t in allset)
        ri    = sum(tpm.get(t, 0) for t in riset)
        if total > 0:
            recs.append({"sample": sample, "gene": gene, "ri_pct": 100 * ri / total})
res = pd.DataFrame(recs)
summary = res.groupby("gene")["ri_pct"].agg(median="median", mean="mean", n="count").reindex(["Klrk1","Actb","Gapdh"])
print("\n=== Retained-intron proportion per gene (negative control) ===")
print(summary.round(2).to_string())
kl, ac, ga = summary.loc["Klrk1","median"], summary.loc["Actb","median"], summary.loc["Gapdh","median"]
print(f"\nKlrk1 RI median {kl:.1f}%  is  {kl/ac:.0f}x Actb ({ac:.2f}%)  and  {kl/ga:.0f}x Gapdh ({ga:.2f}%)")

summary.to_csv(os.path.join(cfg.TABLE_DIR, "housekeeping_negative_control.csv"))
fig, ax = plt.subplots(figsize=(6, 4.5))
order = ["Klrk1", "Actb", "Gapdh"]
ax.bar(order, [summary.loc[g, "median"] for g in order], color=["#c0392b", "#7f8c8d", "#95a5a6"])
ax.set_ylabel("Retained-intron proportion (%), median across samples")
ax.set_title("FigV5 — Housekeeping negative control", fontweight="bold")
for i, g in enumerate(order):
    ax.text(i, summary.loc[g, "median"], f"{summary.loc[g,'median']:.1f}%", ha="center", va="bottom")
fig.tight_layout()
fig.savefig(os.path.join(cfg.FIG_DIR, "FigV5_housekeeping_negative_control.pdf"))
fig.savefig(os.path.join(cfg.FIG_DIR, "FigV5_housekeeping_negative_control.png"), dpi=200)
print(f"\nfigure -> {cfg.FIG_DIR}/FigV5_housekeeping_negative_control.png")
